File: backend/security.py
"""Acces panou: whitelist IP, setări hub, autentificare Telegram WebApp."""

import hashlib
import hmac
import ipaddress
import os
import secrets
from datetime import timedelta
from urllib.parse import parse_qsl

from models import HubSetting, IpWhitelist, TelegramLinkCode, TelegramUser, TelegramWebSession, utcnow

ENV_TELEGRAM_TOKEN = os.environ.get("TELEGRAM_BOT_TOKEN", "")
ENV_TELEGRAM_WEBAPP = os.environ.get("TELEGRAM_WEBAPP_URL", "").rstrip("/")


def get_telegram_bot_token(db=None):
    if db is not None:
        stored = get_setting(db, "telegram_bot_token", "")
        if stored:
            return stored
    return ENV_TELEGRAM_TOKEN


def get_telegram_webapp_url(db=None):
    if db is not None:
        stored = get_setting(db, "telegram_webapp_url", "")
        if stored:
            return stored.rstrip("/")
    return ENV_TELEGRAM_WEBAPP


def mask_token(token):
    if not token:
        return ""
    if len(token) <= 8:
        return "••••"
    return f"••••{token[-6:]}"


def client_ip_from_request(request):
    forwarded = request.headers.get("X-Forwarded-For", "")
    if forwarded:
        return forwarded.split(",")[0].strip()
    real_ip = request.headers.get("X-Real-IP", "")
    if real_ip:
        return real_ip.strip()
    return request.remote_addr or ""


def _ip_matches(allowed, client):
    if not allowed or not client:
        return False
    try:
        if "/" in allowed:
            return ipaddress.ip_address(client) in ipaddress.ip_network(allowed, strict=False)
        return ipaddress.ip_address(client) == ipaddress.ip_address(allowed)
    except ValueError:
        return client == allowed


def is_local_ip(ip):
    return ip in ("127.0.0.1", "::1", "localhost")


def get_setting(db, key, default=""):
    row = db.query(HubSetting).filter_by(key=key).first()
    return row.value if row else default


def set_setting(db, key, value):
    row = db.query(HubSetting).filter_by(key=key).first()
    if not row:
        row = HubSetting(key=key)
        db.add(row)
    row.value = str(value)
    db.commit()


def is_whitelist_enabled(db):
    return get_setting(db, "ip_whitelist_enabled", "0") == "1"


def check_ip_whitelist(db, client_ip):
    if is_local_ip(client_ip):
        return True
    if not is_whitelist_enabled(db):
        return True
    entries = db.query(IpWhitelist).all()
    if not entries:
        return True
    return any(_ip_matches(e.ip, client_ip) for e in entries)


def generate_link_code(db, ttl_minutes=10):
    db.query(TelegramLinkCode).filter(TelegramLinkCode.expires_at < utcnow()).delete()
    db.commit()
    for _ in range(10):
        code = secrets.token_hex(3).upper()[:6]
        if not db.query(TelegramLinkCode).filter_by(code=code, used=False).first():
            break
    expires = utcnow() + timedelta(minutes=ttl_minutes)
    db.add(TelegramLinkCode(code=code, expires_at=expires))
    db.commit()
    return code, expires


def validate_telegram_init_data(init_data, bot_token):
    if not init_data or not bot_token:
        return None
    parsed = dict(parse_qsl(init_data, keep_blank_values=True))
    received_hash = parsed.pop("hash", None)
    if not received_hash:
        return None
    data_check = "\n".join(f"{k}={v}" for k, v in sorted(parsed.items()))
    secret = hmac.new(b"WebAppData", bot_token.encode(), hashlib.sha256).digest()
    calculated = hmac.new(secret, data_check.encode(), hashlib.sha256).hexdigest()
    if calculated != received_hash:
        return None
    import json
    user_raw = parsed.get("user")
    if not user_raw:
        return None
    try:
        return json.loads(user_raw)
    except json.JSONDecodeError:
        return None


def create_telegram_web_session(db, telegram_id, hours=24):
    db.query(TelegramWebSession).filter(TelegramWebSession.expires_at < utcnow()).delete()
    token = secrets.token_urlsafe(32)
    expires = utcnow() + timedelta(hours=hours)
    db.add(TelegramWebSession(token=token, telegram_id=telegram_id, expires_at=expires))
    db.commit()
    return token, expires


def get_telegram_web_session(db, token):
    row = db.query(TelegramWebSession).filter_by(token=token).first()
    if not row or row.expires_at < utcnow():
        return None
    user = db.query(TelegramUser).filter_by(telegram_id=row.telegram_id, is_active=True).first()
    return user

File: hub/src/lib/security.js
import crypto from "node:crypto";
import { randomBytes } from "node:crypto";
import { exec, queryAll, queryOne } from "../db/index.js";
import { config } from "../config.js";

export async function getSetting(key, defaultValue = "") {
  const row = await queryOne("SELECT value FROM hub_settings WHERE `key` = ?", [key]);
  return row ? row.value : defaultValue;
}

export async function setSetting(key, value) {
  const row = await queryOne("SELECT `key` FROM hub_settings WHERE `key` = ?", [key]);
  if (row) await exec("UPDATE hub_settings SET value = ? WHERE `key` = ?", [String(value), key]);
  else await exec("INSERT INTO hub_settings (`key`, value) VALUES (?, ?)", [key, String(value)]);
}

export function maskToken(token) {
  if (!token) return "";
  if (token.length <= 8) return "••••";
  return `••••${token.slice(-6)}`;
}

export function normalizeClientIp(ip) {
  const v = String(ip || "").trim();
  if (v.startsWith("::ffff:")) return v.slice(7);
  return v;
}

export function clientIpFromRequest(req) {
  const xff = req.headers["x-forwarded-for"];
  if (xff) return normalizeClientIp(String(xff).split(",")[0]);
  const xr = req.headers["x-real-ip"];
  if (xr) return normalizeClientIp(xr);
  const cf = req.headers["cf-connecting-ip"];
  if (cf) return normalizeClientIp(cf);
  return normalizeClientIp(req.ip || req.socket?.remoteAddress || "");
}

function ipv4ToInt(ip) {
  const p = ip.split(".").map((x) => parseInt(x, 10));
  if (p.length !== 4 || p.some((n) => Number.isNaN(n) || n < 0 || n > 255)) throw new Error("bad ipv4");
  return ((p[0] << 24) | (p[1] << 16) | (p[2] << 8) | p[3]) >>> 0;
}

function ipMatch(allowed, client) {
  if (!allowed || !client) return false;
  const a = normalizeClientIp(allowed.trim());
  const c = normalizeClientIp(client.trim());
  if (!a.includes("/")) return a === c;
  const [base, bitsStr] = a.split("/");
  const bits = parseInt(bitsStr, 10);
  if (Number.isNaN(bits) || bits < 0 || bits > 32) return false;
  try {
    const mask = bits === 0 ? 0 : (~0 << (32 - bits)) >>> 0;
    return (ipv4ToInt(base) & mask) === (ipv4ToInt(c) & mask);
  } catch {
    return false;
  }
}

export async function isWhitelistEnabled() {
  return (await getSetting("ip_whitelist_enabled", "0")) === "1";
}

export async function checkIpWhitelist(clientIp) {
  if (!(await isWhitelistEnabled())) return true;
  const ip = normalizeClientIp(clientIp);
  const entries = await queryAll("SELECT ip FROM ip_whitelist", []);
  if (!entries.length) return false;
  if (config.panelAllowLocalhost && ["127.0.0.1", "::1", "localhost"].includes(ip)) return true;
  return entries.some((e) => ipMatch(e.ip, ip));
}

export async function generateLinkCode(ttlMinutes = 10) {
  await exec("DELETE FROM telegram_link_codes WHERE expires_at < ?", [new Date().toISOString()]);
  let code = "";
  for (let i = 0; i < 10; i += 1) {
    code = randomBytes(3).toString("hex").toUpperCase().slice(0, 6);
    const row = await queryOne("SELECT code FROM telegram_link_codes WHERE code = ? AND used = 0", [code]);
    if (!row) break;
  }
  const expires = new Date(Date.now() + ttlMinutes * 60 * 1000).toISOString();
  await exec("INSERT INTO telegram_link_codes (code, created_at, expires_at, used) VALUES (?, ?, ?, 0)", [code, new Date().toISOString(), expires]);
  return { code, expires };
}

export function validateTelegramInitData(initData, botToken) {
  if (!initData || !botToken) return null;
  const params = new URLSearchParams(initData);
  const receivedHash = params.get("hash");
  if (!receivedHash) return null;
  params.delete("hash");
  const entries = [...params.entries()].sort(([a], [b]) => a.localeCompare(b));
  const dataCheck = entries.map(([k, v]) => `${k}=${v}`).join("\n");
  const secret = crypto.createHmac("sha256", "WebAppData").update(botToken).digest();
  const calc = crypto.createHmac("sha256", secret).update(dataCheck).digest("hex");
  if (calc !== receivedHash) return null;
  try {
    return JSON.parse(params.get("user") || "");
  } catch {
    return null;
  }
}

export async function createTelegramWebSession(telegramId, hours = 24) {
  await exec("DELETE FROM telegram_web_sessions WHERE expires_at < ?", [new Date().toISOString()]);
  const token = randomBytes(32).toString("base64url");
  const expires = new Date(Date.now() + hours * 3600 * 1000).toISOString();
  await exec("INSERT INTO telegram_web_sessions (token, telegram_id, expires_at, created_at) VALUES (?, ?, ?, ?)", [
    token, telegramId, expires, new Date().toISOString()
  ]);
  return { token, expires };
}

export async function getTelegramWebSession(token) {
  const row = await queryOne("SELECT * FROM telegram_web_sessions WHERE token = ? AND expires_at >= ?", [token, new Date().toISOString()]);
  if (!row) return null;
  return queryOne("SELECT * FROM telegram_users WHERE telegram_id = ? AND is_active = 1", [row.telegram_id]);
}

export async function getTelegramBotToken() {
  return (await getSetting("telegram_bot_token", "")) || config.telegramBotToken;
}

export async function getTelegramWebAppUrl() {
  const v = (await getSetting("telegram_webapp_url", "")) || config.telegramWebAppUrl;
  return String(v).replace(/\/+$/, "");
}
